[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x157d737]

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers.

To fix problems, we need clear reproduction cases - we need to be able to see it happen locally. A reproduction case is ideally something a Terraform Core engineer can git-clone or copy-paste and run immediately, without inventing any details or context. 

* A short example can be directly copy-pasteable; longer examples should be in separate git repositories, especially if multiple files are needed
* Please include all needed context. For example, if you figured out that an expression can cause a crash, put the expression in a variable definition or a resource
* Set defaults on (or omit) any variables. The person reproducing it should not need to invent variable settings
* If multiple steps are required, such as running terraform twice, consider scripting it in a simple shell script. For example, see [this case](https://github.com/danieldreier/terraform-issue-reproductions/tree/master/25719). Providing a script can be easier than explaining what changes to make to the config between runs.
* Omit any unneeded complexity: remove variables, conditional statements, functions, modules, providers, and resources that are not needed to trigger the bug
* When possible, use the [null resource](https://www.terraform.io/docs/providers/null/resource.html) provider rather than a real provider in order to minimize external dependencies. We know this isn't always feasible. The Terraform Core team doesn't have deep domain knowledge in every provider, or access to every cloud platform for reproduction cases.

-->

### Terraform Version
<!---
Run `terraform version` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.15.4
on darwin_amd64
+ provider registry.terraform.io/hashicorp/aws v3.42.0
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/hashicorp/random v3.1.0
+ provider registry.terraform.io/hashicorp/template v2.2.0
+ provider registry.terraform.io/hashicorp/vault v2.20.0
+ provider registry.terraform.io/nolte/awx v0.2.2
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For Terraform configs larger than a few resources, or that involve multiple files, please make a GitHub repository that we can clone, rather than copy-pasting multiple files in here. For security, you can also encrypt the files using our GPG public key at https://www.hashicorp.com/security.
-->

```terraform
terraform {
  backend "s3" {}
}

provider "vault" {
  address = var.VAULT_ADDR
}

data "vault_generic_secret" "aws" {
  path = var.vault_path_aws_keys
}

provider "aws" {
  region     = var.aws_region
  access_key = data.vault_generic_secret.aws.data["access_key"]
  secret_key = data.vault_generic_secret.aws.data["secret_key"]
}

data "aws_vpc" "main" {
  filter {
    name   = "tag:Name"
    values = [var.vpc-name]
  }
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
```
2021-05-24T14:55:50.371+0100 [TRACE] Re-validating config for "aws_route53_record.dns[\"DOMAIN1\"]"
2021-05-24T14:55:50.371+0100 [TRACE] GRPCProvider: ValidateResourceConfig
2021-05-24T14:55:50.371+0100 [INFO]  provider.terraform-provider-aws_v3.42.0_x5: 2021/05/24 14:55:50 [WARN] Truncating attribute path of 0 diagnostics for TypeSet: timestamp=2021-05-24T14:55:50.371+0100
2021-05-24T14:55:50.372+0100 [TRACE] GRPCProvider: PlanResourceChange
2021-05-24T14:55:50.373+0100 [WARN]  Provider "registry.terraform.io/hashicorp/aws" produced an invalid plan for aws_route53_record.dns["DOMAIN1"], but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .set_identifier: planned value cty.StringVal("") for a non-computed attribute
      - .ttl: planned value cty.NumberIntVal(0) for a non-computed attribute
      - .health_check_id: planned value cty.StringVal("") for a non-computed attribute
      - .records: planned value cty.SetValEmpty(cty.String) for a non-computed attribute
2021-05-24T14:55:50.373+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for aws_route53_record.dns["DOMAIN1"]
2021-05-24T14:55:50.373+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for aws_route53_record.dns["DOMAIN1"]
2021-05-24T14:55:50.373+0100 [TRACE] writeChange: recorded NoOp change for aws_route53_record.dns["DOMAIN1"]
2021-05-24T14:55:50.373+0100 [TRACE] vertex "aws_route53_record.dns[\"DOMAIN1\"]": visit complete
2021-05-24T14:55:50.423+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to refreshState for aws_route53_record.dns["DOMAIN2"]
2021-05-24T14:55:50.423+0100 [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: writing state object for aws_route53_record.dns["DOMAIN2"]
2021-05-24T14:55:50.424+0100 [TRACE] Re-validating config for "aws_route53_record.dns[\"DOMAIN2\"]"
2021-05-24T14:55:50.424+0100 [TRACE] GRPCProvider: ValidateResourceConfig
2021-05-24T14:55:50.424+0100 [INFO]  provider.terraform-provider-aws_v3.42.0_x5: 2021/05/24 14:55:50 [WARN] Truncating attribute path of 0 diagnostics for TypeSet: timestamp=2021-05-24T14:55:50.424+0100
2021-05-24T14:55:50.424+0100 [TRACE] GRPCProvider: PlanResourceChange
2021-05-24T14:55:50.425+0100 [WARN]  Provider "registry.terraform.io/hashicorp/aws" produced an invalid plan for aws_route53_record.dns["DOMAIN2"], but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .ttl: planned value cty.NumberIntVal(0) for a non-computed attribute
      - .health_check_id: planned value cty.StringVal("") for a non-computed attribute
      - .records: planned value cty.SetValEmpty(cty.String) for a non-computed attribute
      - .set_identifier: planned value cty.StringVal("") for a non-computed attribute
XXX
2021-05-24T14:55:50.615+0100 [TRACE] vertex "meta.count-boundary (EachMode fixup)": starting visit (*terraform.NodeCountBoundary)
2021-05-24T14:55:50.615+0100 [TRACE] vertex "provider[\"registry.terraform.io/hashicorp/aws\"] (close)": starting visit (*terraform.graphNodeCloseProvider)
2021-05-24T14:55:50.615+0100 [TRACE] GRPCProvider: Close
2021-05-24T14:55:50.615+0100 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = transport is closing"
2021-05-24T14:55:50.616+0100 [TRACE] vertex "meta.count-boundary (EachMode fixup)": visit complete
2021-05-24T14:55:50.621+0100 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/hashicorp/aws/3.42.0/darwin_amd64/terraform-provider-aws_v3.42.0_x5 pid=72967
2021-05-24T14:55:50.621+0100 [DEBUG] provider: plugin exited
2021-05-24T14:55:50.621+0100 [TRACE] vertex "provider[\"registry.terraform.io/hashicorp/aws\"] (close)": visit complete
2021-05-24T14:55:50.621+0100 [TRACE] vertex "root": starting visit (*terraform.nodeCloseModule)
2021-05-24T14:55:50.621+0100 [TRACE] vertex "root": visit complete

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x157d737]

goroutine 65 [running]:
github.com/hashicorp/terraform/internal/states.(*ResourceInstanceObjectSrc).Decode(0x0, 0x38b7cb8, 0xc001f46300, 0xc003904690, 0x0, 0x0)
	/home/circleci/project/project/internal/states/instance_object_src.go:77 +0x37
github.com/hashicorp/terraform/internal/command/format.ResourceInstanceDrift(0xc000ba3680, 0x2, 0x2, 0x4d, 0xc00072be78, 0x18, 0xc000c796c0, 0x4, 0x38a9dd8, 0x484e100, ...)
	/home/circleci/project/project/internal/command/format/diff.go:224 +0x465
github.com/hashicorp/terraform/internal/command/views.renderChangesDetectedByRefresh(0xc0012c8a90, 0xc0012c8a98, 0xc0011019e0, 0xc0001bf300, 0xc0009fe340)
	/home/circleci/project/project/internal/command/views/plan.go:382 +0x674
github.com/hashicorp/terraform/internal/command/views.renderPlan(0xc001f68f00, 0xc0011019e0, 0xc0001bf300)
	/home/circleci/project/project/internal/command/views/plan.go:100 +0x70
github.com/hashicorp/terraform/internal/command/views.(*OperationHuman).Plan(0xc000a84920, 0xc001f68f00, 0xc0011019e0)
	/home/circleci/project/project/internal/command/views/operation.go:89 +0x45
github.com/hashicorp/terraform/internal/backend/local.(*Local).opApply(0xc000c200a0, 0x38b5ef8, 0xc000b900c0, 0x38b5ef8, 0xc000b90100, 0xc000652240, 0xc000b90000)
	/home/circleci/project/project/internal/backend/local/backend_apply.go:78 +0xedd
github.com/hashicorp/terraform/internal/backend/local.(*Local).Operation.func1(0xc000a84ed0, 0xc000a84ee0, 0xc000a84ef0, 0xc000c200a0, 0xc000a84ec0, 0x38b5ef8, 0xc000b900c0, 0x38b5ef8, 0xc000b90100, 0xc000652240, ...)
	/home/circleci/project/project/internal/backend/local/backend.go:321 +0xe8
created by github.com/hashicorp/terraform/internal/backend/local.(*Local).Operation
	/home/circleci/project/project/internal/backend/local/backend.go:315 +0x2f2
```

### Expected Behavior
for DNSs to be created

### Actual Behavior
crash

### Steps to Reproduce
terraform apply -compact-warnings  -parallelism=300 -var-file=global-dev.tfvars -var-file=dev.tfvars

--- a/internal/command/format/diff.go
+++ b/internal/command/format/diff.go
@@ -172,6 +172,12 @@ func ResourceInstanceDrift(
 	action := plans.Update
 
 	switch {
+	case before == nil || before.Current == nil:
+		// before should never be nil, but before.Current can be if the
+		// instance was deposed. There is nothing to render for a deposed
+		// instance, since we intend to remove it.
+		return ""
+
 	case after == nil || after.Current == nil:
 		// The object was deleted
 		buf.WriteString(color.Color(fmt.Sprintf("[bold]  # %s[reset] has been deleted", dispAddr)))

--- a/internal/command/views/operation_test.go
+++ b/internal/command/views/operation_test.go
@@ -112,6 +112,33 @@ func TestOperation_planNoChanges(t *testing.T) {
 			},
 			"No objects need to be destroyed.",
 		},
+		"no drift to display with only deposed instances": {
+			// changes in deposed instances will cause a change in state, but
+			// have nothing to display to the user
+			func(schemas *terraform.Schemas) *plans.Plan {
+				return &plans.Plan{
+					UIMode:  plans.NormalMode,
+					Changes: plans.NewChanges(),
+					PrevRunState: states.BuildState(func(state *states.SyncState) {
+						state.SetResourceInstanceDeposed(
+							addrs.Resource{
+								Mode: addrs.ManagedResourceMode,
+								Type: "test_resource",
+								Name: "somewhere",
+							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
+							states.NewDeposedKey(),
+							&states.ResourceInstanceObjectSrc{
+								Status:    states.ObjectReady,
+								AttrsJSON: []byte(`{"foo": "ok", "bars":[]}`),
+							},
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
+						)
+					}),
+					PriorState: states.NewState(),
+				}
+			},
+			"no differences, so no changes are needed.",
+		},
 		"drift detected in normal mode": {
 			func(schemas *terraform.Schemas) *plans.Plan {
 				return &plans.Plan{

--- a/internal/command/views/operation_test.go
+++ b/internal/command/views/operation_test.go
@@ -121,17 +121,80 @@ func TestOperation_planNoChanges(t *testing.T) {
 						state.SetResourceInstanceCurrent(
 							addrs.Resource{
 								Mode: addrs.ManagedResourceMode,
-								Type: "something",
+								Type: "test_resource",
 								Name: "somewhere",
 							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
 							&states.ResourceInstanceObjectSrc{
 								Status:    states.ObjectReady,
 								AttrsJSON: []byte(`{}`),
 							},
-							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewBuiltInProvider("test")),
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
 						)
 					}),
 					PriorState: states.NewState(),
+				}
+			},
+			"to update the Terraform state to match, create and apply a refresh-only plan",
+		},
+		"drift detected with deposed": {
+			func(schemas *terraform.Schemas) *plans.Plan {
+				return &plans.Plan{
+					UIMode:  plans.NormalMode,
+					Changes: plans.NewChanges(),
+					PrevRunState: states.BuildState(func(state *states.SyncState) {
+						state.SetResourceInstanceCurrent(
+							addrs.Resource{
+								Mode: addrs.ManagedResourceMode,
+								Type: "test_resource",
+								Name: "changes",
+							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
+							&states.ResourceInstanceObjectSrc{
+								Status:    states.ObjectReady,
+								AttrsJSON: []byte(`{"foo":"b"}`),
+							},
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
+						)
+						state.SetResourceInstanceDeposed(
+							addrs.Resource{
+								Mode: addrs.ManagedResourceMode,
+								Type: "test_resource",
+								Name: "broken",
+							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
+							states.NewDeposedKey(),
+							&states.ResourceInstanceObjectSrc{
+								Status:    states.ObjectReady,
+								AttrsJSON: []byte(`{"foo":"c"}`),
+							},
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
+						)
+					}),
+					PriorState: states.BuildState(func(state *states.SyncState) {
+						state.SetResourceInstanceCurrent(
+							addrs.Resource{
+								Mode: addrs.ManagedResourceMode,
+								Type: "test_resource",
+								Name: "changed",
+							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
+							&states.ResourceInstanceObjectSrc{
+								Status:    states.ObjectReady,
+								AttrsJSON: []byte(`{"foo":"b"}`),
+							},
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
+						)
+						state.SetResourceInstanceDeposed(
+							addrs.Resource{
+								Mode: addrs.ManagedResourceMode,
+								Type: "test_resource",
+								Name: "broken",
+							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
+							states.NewDeposedKey(),
+							&states.ResourceInstanceObjectSrc{
+								Status:    states.ObjectReady,
+								AttrsJSON: []byte(`{"foo":"d"}`),
+							},
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
+						)
+					}),
 				}
 			},
 			"to update the Terraform state to match, create and apply a refresh-only plan",
@@ -145,14 +208,14 @@ func TestOperation_planNoChanges(t *testing.T) {
 						state.SetResourceInstanceCurrent(
 							addrs.Resource{
 								Mode: addrs.ManagedResourceMode,
-								Type: "something",
+								Type: "test_resource",
 								Name: "somewhere",
 							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
 							&states.ResourceInstanceObjectSrc{
 								Status:    states.ObjectReady,
 								AttrsJSON: []byte(`{}`),
 							},
-							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewBuiltInProvider("test")),
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
 						)
 					}),
 					PriorState: states.NewState(),
@@ -169,14 +232,14 @@ func TestOperation_planNoChanges(t *testing.T) {
 						state.SetResourceInstanceCurrent(
 							addrs.Resource{
 								Mode: addrs.ManagedResourceMode,
-								Type: "something",
+								Type: "test_resource",
 								Name: "somewhere",
 							}.Instance(addrs.NoKey).Absolute(addrs.RootModuleInstance),
 							&states.ResourceInstanceObjectSrc{
 								Status:    states.ObjectReady,
 								AttrsJSON: []byte(`{}`),
 							},
-							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewBuiltInProvider("test")),
+							addrs.RootModuleInstance.ProviderConfigDefault(addrs.NewDefaultProvider("test")),
 						)
 					}),
 					PriorState: states.NewState(),

--- a/internal/command/views/plan.go
+++ b/internal/command/views/plan.go
@@ -339,7 +339,7 @@ func renderPlan(plan *plans.Plan, schemas *terraform.Schemas, view *View) {
 // line of output, and guarantees to always produce whole lines terminated
 // by newline characters.
 func renderChangesDetectedByRefresh(before, after *states.State, schemas *terraform.Schemas, view *View) bool {
-	// ManagedResourceEqual checks that the state es exactly equal for all
+	// ManagedResourceEqual checks that the state is exactly equal for all
 	// managed resources; but semantically equivalent states, or changes to
 	// deposed instances may not actually represent changes we need to present
 	// to the user, so for now this only serves as a short-circuit to skip

--- a/internal/command/views/plan.go
+++ b/internal/command/views/plan.go
@@ -374,6 +374,10 @@ func renderChangesDetectedByRefresh(before, after *states.State, schemas *terraf
 			}
 
 			for key, bis := range brs.Instances {
+				if bis.Current == nil {
+					// No current instance to render here
+					continue
+				}
 				var pis *states.ResourceInstance
 				if prs != nil {
 					pis = prs.Instance(key)

--- a/internal/command/views/plan.go
+++ b/internal/command/views/plan.go
@@ -339,17 +339,16 @@ func renderPlan(plan *plans.Plan, schemas *terraform.Schemas, view *View) {
 // line of output, and guarantees to always produce whole lines terminated
 // by newline characters.
 func renderChangesDetectedByRefresh(before, after *states.State, schemas *terraform.Schemas, view *View) bool {
+	// ManagedResourceEqual checks that the state es exactly equal for all
+	// managed resources; but semantically equivalent states, or changes to
+	// deposed instances may not actually represent changes we need to present
+	// to the user, so for now this only serves as a short-circuit to skip
+	// attempting to render the diffs below.
 	if after.ManagedResourcesEqual(before) {
 		return false
 	}
 
-	view.streams.Print(
-		view.colorize.Color("[reset]\n[bold][cyan]Note:[reset][bold] Objects have changed outside of Terraform[reset]\n\n"),
-	)
-	view.streams.Print(format.WordWrap(
-		"Terraform detected the following changes made outside of Terraform since the last \"terraform apply\":\n\n",
-		view.outputColumns(),
-	))
+	var diffs []string
 
 	for _, bms := range before.Modules {
 		for _, brs := range bms.Resources {
@@ -390,13 +389,31 @@ func renderChangesDetectedByRefresh(before, after *states.State, schemas *terraf
 					view.colorize,
 				)
 				if diff != "" {
-					view.streams.Print(diff)
+					diffs = append(diffs, diff)
 				}
 			}
 		}
 	}
 
-	return true
+	// If we only have changes regarding deposed instances, or the diff
+	// renderer is suppressing irrelevant changes from the legacy SDK, there
+	// may not have been anything to display to the user.
+	if len(diffs) > 0 {
+		view.streams.Print(
+			view.colorize.Color("[reset]\n[bold][cyan]Note:[reset][bold] Objects have changed outside of Terraform[reset]\n\n"),
+		)
+		view.streams.Print(format.WordWrap(
+			"Terraform detected the following changes made outside of Terraform since the last \"terraform apply\":\n\n",
+			view.outputColumns(),
+		))
+
+		for _, diff := range diffs {
+			view.streams.Print(diff)
+		}
+		return true
+	}
+
+	return false
 }
 
 const planHeaderIntro = `

